src/Nethermind/Nethermind.JsonRpc/Modules/Subscribe/LogsSubscription.cs

```diff
@@ -0,0 +1,127 @@
+using Nethermind.Blockchain.Receipts;
+using Nethermind.Core;
+using Nethermind.JsonRpc.Modules.Eth;
+using Nethermind.Logging;
+
+namespace Nethermind.JsonRpc.Modules.Subscribe
+{
+    public class LogsSubscription : Subscription
+    {
+        private readonly IReceiptStorage _receiptStorage;
+        private readonly IBlockFinder _blockFinder;
+        private readonly LogFilter _filter;
+        private readonly ILogger _logger;
+
+        public LogsSubscription(IReceiptStorage receiptStorage, IFilterStore store, IBlockFinder blockFinder, ILogManager logManager, Filter filter = null)
+        {
+            _receiptStorage = receiptStorage;
+            _blockFinder = blockFinder;
+            _logger = logManager.GetClassLogger();
+            IFilterStore filterStore = store;
+
+            if (filter != null)
+            {
+                _filter = filterStore.CreateLogFilter(
+                    filter.FromBlock,
+                    filter.ToBlock,
+                    filter.Address,
+                    filter.Topics);
+                if (_logger.IsTrace) _logger.Trace($"Logs Subscription {Id}: Created LogFilter with the same arguments like \"filter\"");
+            }
+            else
+            {
+                _filter = filterStore.CreateLogFilter(
+                    BlockParameter.Latest,
+                    BlockParameter.Latest);
+                if (_logger.IsTrace) _logger.Trace($"Logs Subscription {Id}: Argument \"filter\" was null and created LogFilter with arguments: FromBlock: BlockParameter.Latest, ToBlock: BlockParameter.Latest");
+            }
+            
+            _receiptStorage.ReceiptsInserted += OnReceiptsInserted;
+            if(_logger.IsTrace) _logger.Trace($"Logs subscription {Id} will track ReceiptsInserted.");
+        }
+
+        private void OnReceiptsInserted(object? sender, ReceiptsEventArgs e)
+        {
+            Task.Run(() =>
+            {
+                BlockHeader fromBlock = _blockFinder.FindHeader(_filter.FromBlock);
+                BlockHeader toBlock = _blockFinder.FindHeader(_filter.ToBlock, true);
+
+                if (e.BlockHeader.Number >= fromBlock.Number
+                    && e.BlockHeader.Number <= toBlock.Number)
+                {
+                    var filterLogsArguments = e.BlockHeader.FindLogsWithReceipts(e.TxReceipts, _filter);
+                    
+                    foreach (var filterLogArguments in filterLogsArguments)
+                    {
+                        FilterLog filterLog = new FilterLog(
+                            filterLogArguments.LogIndex,
+                            filterLogArguments.TransactionLogIndex,
+                            filterLogArguments.Receipt,
+                            filterLogArguments.Entry);
+                        
+                        JsonRpcResult result = GetJsonRpcResult(filterLog);
+                        JsonRpcDuplexClient.SendJsonRpcResult(result);
```

Comment:
In general in all subscriptions JsonRpcDuplexClient can be null before bound to subscription for short period of time. We should either check that before invoking Send or maybe even better only attach to events when its set to not-null

Reply:
nullable references could potentially issue a warning then